### PR TITLE
Explosions Part I - Directional Explosions

### PR DESCRIPTION
**Pull request by Thunder12345**
## About The Pull Request

Adds the ability for explosions to be directional. This is achieved by adding an angle check to `prepare_explosion_turfs()` to drop any turfs outside the cone of the explosion. If the arc covers a full 360 degrees, as is the default, it will accept all the turfs without performing the angle check.

Uses this functionality to rework both rocket launcher backblast and X4 explosions. Rocket launcher backblast has been changed from a shotgun of indendiary bullets to a directional explosion of similar length. X4 now uses a directional explosion to "ensure user safety".

Apparently the old method of moving the explosion one tile away didn't even work, as it blew up `target` before trying to check its density for the directional behaviour.

https://youtu.be/Mzdt7d7Le2Y

## Why It's Good For The Game

Directional explosions - Useful functionality for a range of potential use cases, which can be implemented with minimal extra processing cost (Worst case scenario being very large directional explosions)

Backblast - Looks way cooler than a bunch of projectiles, and should be significantly more functional in high-lag situations where projectile code tends to get fucky

X4 - More predictable for players wanting to use it as a breaching charge, you can actually stand near the charge and not have to worry about being hoist upon your own petard.

## Changelog
:cl:
add: Added support for directional explosions.
add: Rocket launcher backblast is now 271% more explosive, check your six for friendlies!
add: X4 charges now explode in a cone away from the user when placed on a sufficiently solid object.
fix: X4 charges will now behave correctly when placed on dense atoms (note: don't try to read a variable from an atom you just blew up)
/:cl:

### CODE REVIEW

File: code/__DEFINES/explosions.dm [hunk 1]
```diff
@@ -52,6 +52,12 @@
 #define EXARG_KEY_SILENT STRINGIFY(silent)
 /// Whether or not the explosion should produce smoke if it is large enough to warrant it.
 #define EXARG_KEY_SMOKE STRINGIFY(smoke)
+///
```
Comment (Rhials) [inline code/__DEFINES/explosions.dm:55]: You're missing a comment!!!!!!

**Comment by Paxilmaniac** [comment]
Directional explosions sounds like I can get up to a lot of trolling with rocket launcher code, this will be bad for the space carp season.

**Comment by Dawnseer** [comment]
Now here's hoping someone will bring this new advanced directional technology to pipebombs, so I can finally make my IED with a lovely 'this side towards tiders' stickers slapped on

**Review by Time-Green — CHANGES_REQUESTED** [review]
Nice work! Some small suggestions to help with clarity

File: code/controllers/subsystem/explosions.dm [hunk 2]
```diff
@@ -580,43 +582,77 @@ SUBSYSTEM_DEF(explosions)
...
 
 	var/our_x = epicenter.x
 	var/our_y = epicenter.y
 	var/our_z = epicenter.z
 
 	var/max_x = world.maxx
 	var/max_y = world.maxy
+
+	// Work out the angles to explode between
+	var/first_angle_limit = WRAP(explosion_direction - explosion_arc/2, 0, 360)
```
[the remark this answers is not in the document]
Comment (Time-Green) [reply code/controllers/subsystem/explosions.dm:600]: I forgot to explain but multiplication is just faster in byond

**Comment by DrDiasyl** [comment]
We now need to give Nukies claymores so they can have true CoD experience

**Comment by carlarctg** [comment]
Nominative Determinism strikes again

**Comment by deathrobotpunch** [comment]
TG: THE CLAYMORE UPDATE

**Pull request merged**